File: 3747. Count Distinct Integers After Removing Zeros-1.py
class Solution:
    def countDistinct(self, n: int) -> int:
        s = str(n)
        m = len(s)

        pow9 = 9**m
        res = (pow9 - 9) // 8

        for i, d in enumerate(s):
            if d == "0":
                break
            v = int(d) - 1
            if i == m - 1:
                v += 1
            pow9 //= 9
            res += v * pow9

        return res

File: 3747. Count Distinct Integers After Removing Zeros-1.cpp
// mark
// 1848
// 數位 + 組合數學
class Solution {
public:
  long long countDistinct(long long n) {
    string s = "";
    while (n) {
      int d = n % 10;
      s = char(d + '0') + s;
      n /= 10;
    }

    int m = s.size();
    long long res = 0, pow9 = 1;
    for (int i = 1; i <= m - 1; ++i) {
      pow9 *= 9;
      res += pow9;
    }

    for (int i = 0; i < m; ++i) {
      if (s[i] == '0')
        break;
      int v = s[i] - '0' - 1;
      if (i == m - 1)
        v++;
      res += v * pow9;
      pow9 /= 9;
    }

    return res;
  }
};